WRQ-9660: Updated samples to use `useReducer` rather than redux

mmyelyn opened the pull request:
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
There are a total of 4 samples that use redux related libraries inside sandstone's samples.
That's qa-virtuallist, qa-virtualgridlist, qa-virtualgridlistnative-preserve-focus, and event-logger.
I have deleted redux related libraries to avoid increasing the number of libraries that need to be managed within the sample.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
The store implementation using redux was deleted and changed to the context and useReducer implementation.
refer to : https://react.dev/learn/scaling-up-with-reducer-and-context

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)


### Links
[//]: # (Related issues, references)
WRQ-9660

### Comments
Enact-DCO-1.0-Signed-off-by: Hyelyn Kim ([myelyn.kim@lge.com](mailto:sc.baek@lge.com))

juwonjeong commented:
Please merge the recent develop branch to solve the validate-docs error in travis.

juwonjeong commented:
Did you check the lint warnings? Please fix all of it. You can also check it locally using the `enact lint --strict` command.

juwonjeong commented on samples/event-logger/src/components/InputBoard/InputBoard.js:
```diff
@@ -64,15 +82,21 @@ const InputBoard = ({className}) => {
-	const onUpdateEventLog = useCallback((prevTimeoutId, postTimeoutId, eventObject) => dispatch(updateEventLog(prevTimeoutId, postTimeoutId, eventObject)), [dispatch]);
+	const onAddEventLog =
+		useCallback(
+			(timeoutId, eventName, isDOMElement, isCapturing, eventObject) =>
```
Please check the order of the parameters

mmyelyn replied:
Modified to match the parameter order of the action function.

juwonjeong commented:
@mmyelyn, Please check console warning in qa-virtualgridlist sample. Also, scrolling doesn't seem to work in this sample.

mmyelyn commented:
> @mmyelyn, Please check console warning in qa-virtualgridlist sample. Also, scrolling doesn't seem to work in this sample.

Fixed console warning and scrolling in qa-virtualgridlist. and also i found a timeout issue in event-logger. So I fixed that, Please refer to the link for more details about that issue.

juwonjeong commented:
@mmyelyn, I found one issue in qa-virtualgridlist sample. Could you please check it? 
When I set dataSize to 10, and change size with +, - buttons, the number of the dataSize is not changed and fixed at 10.
However, in the original sample, the number of the dataSize is updated when changes size with +, - buttons.

mmyelyn commented:
@juwonjeong Sorry, It's fixed. Also, I was managing the value twice using 'useState', but I deleted it as it seemed unnecessary. In the process, I found missing cases and added to the reducer.

juwonjeong reviewed (approved):
LGTM

MikyungKim commented:
@mmyelyn , Please update your branch with the latest develop and fix the merge conflicts.

MikyungKim commented on samples/qa-virtualgridlist/src/components/ImageItem/ImageItem.js:
```diff
@@ -1,29 +1,28 @@
 import ImageItem from '@enact/sandstone/ImageItem';
 
-import {selectItem} from '../../store';
+import {RecordContext, RecordDispatchContext} from '../../context/RecordContext';
+import {selectItem as selectItemAction} from '../../context/RecordContext';
```
Let's make this one line.

MikyungKim commented on samples/qa-virtualgridlist/src/components/ImageItem/ImageItem.js:
```diff
@@ -1,29 +1,28 @@
-		onClick: () => dispatch(selectItem(dataIndex))
-	};
+	const onClick = () => (dispatch(selectItemAction(dataIndex)));
+	return (
```
```suggestion

	return (
```

MikyungKim commented on samples/qa-virtualgridlist/src/components/ImageItem/ImageItem.js:
```diff
@@ -1,29 +1,28 @@
 
-export default connect(mapStateToProps, mapDispatchToProps)(ImageItem);
+export default ImageItemComponent;
+
```
```suggestion
```

mmyelyn replied:
Isn't the rule to leave a single line space at last line?

MikyungKim replied:
That's correct. But it had 2 empty lines at the end.

MikyungKim commented on samples/qa-virtualgridlist/src/components/ImageItem/ImageItem.js:
```diff
@@ -1,29 +1,28 @@
 };
 
-export default connect(mapStateToProps, mapDispatchToProps)(ImageItem);
+export default ImageItemComponent;
```
The exported component's name and the file name is not matching.
Would you either rename this file or use another name for sandstone ImageItem.

mmyelyn replied:
I renamed the file name. I want to keep sandstone ImageItem name

MikyungKim commented on samples/qa-virtualgridlist/src/components/PanelHeader/PanelHeader.js:
```diff
@@ -22,202 +21,117 @@ import {
 
-const PanelHeader = kind({
-	name: 'PanelHeader',
+import createRecord from '../../utils';
```
Let's remove line 28.

MikyungKim commented on samples/qa-virtualgridlist/src/context/RecordContext.js:
```diff
@@ -0,0 +1,168 @@
+import {createContext, useReducer} from "react";
+import {initializeRecords} from "../utils";
```
```suggestion
import {createContext, useReducer} from 'react';

import {initializeRecords} from ../utils';
```

MikyungKim commented on samples/qa-virtualgridlist/src/context/RecordContext.js:
```diff
@@ -0,0 +1,168 @@
+};
+
+/* action type */
+const ADD_ITEM = 'recordReducer/ADDITEM';
+const CHANGE_DATASIZE = 'recordReducer/CHANGEDATASIZE';
+const CHANGE_MIN_HEIGHT = 'recordReducer/CHANGEMINHEIGHT';
+const CHANGE_MIN_WIDTH = 'recordReducer/CHANGEMINWIDTH';
+const CHANGE_SPACING = 'recordReducer/CHANGESPACING';
+const DELETE_ITEM = 'recordReducer/DELETEITEM';
+const DELETE_SELECTED_ITEM = 'recordReducer/DELETESELECTEDITEM';
+const SELECT_ALL = 'recordReducer/SELECTALL';
+const SELECT_ITEM = 'recordReducer/SELECTITEM';
+const SELECTION_ENABLE = 'recordReducer/SELECTIONENABLE';
+const SET_DATA = 'recordReducer/SETDATA';
```
Is there a guide or rule to make these action type name?
If not, I prefer the same string as the variable name like `ADD_ITEM`.

mmyelyn replied:
Redux uses uppercase camel case, and users seem to use that convention as a convention. and It seems React uses it in another way. So, it's up to us to decide, but i follow conventional rules anyway

https://redux.js.org/style-guide/#write-action-types-as-domaineventname
https://react.dev/reference/react/useReducer

MikyungKim replied:
👍 Let's figure out which case is dominant recently.

MikyungKim replied:
We decided to go as is.

MikyungKim commented on samples/qa-virtualgridlist/src/context/RecordContext.js:
```diff
@@ -0,0 +1,168 @@
+export default function recordReducer (state, action) {
+	switch (action.type) {
+		case (ADD_ITEM): {
+			const
+				addedKey = Object.keys(state.data).length,
+				newData = Object.assign({}, state.data);
```
Let's separate these 2 lines to each const define statement.
Please do the same for other ones in this file.

MikyungKim commented on samples/event-logger/src/components/InputBoard/Filter/Filter.js:
```diff
@@ -1,29 +1,32 @@
+import {setEventCapturing as setEventCapturingAction} from '../../../reducer/eventCapturingOnReducer';
+import {isSyntheticEventOn as isSyntheticEventOnAction} from '../../../reducer/syntheticEventOnReducer';
+import {setTimerIndex as setTimerIndexAction} from '../../../reducer/timerIndexReducer';
+
```
```suggestion
```

MikyungKim commented on samples/qa-virtuallist/src/components/ListItem/ListItem.js:
```diff
@@ -1,28 +1,24 @@
-});
+const ListItem = (props) => {
+	const {children, index, ...rest} = props;
+	const {listItems} = useContext(ListContext);
+	const disabled = listItems[index].disabled;
```
```suggestion
	const {disabled} = useContext(ListContext).listItems[index];
```

MikyungKim commented on samples/qa-virtuallist/src/context/ListContext.js:
```diff
@@ -0,0 +1,40 @@
+import {createContext, useReducer} from "react";
```
```suggestion
import {createContext, useReducer} from 'react';
```

MikyungKim commented on samples/qa-virtuallist/src/context/ListContext.js:
```diff
@@ -0,0 +1,40 @@
+const initialData = {listItems: []};
+
+export const ListProvider = ({children}) => {
+
```
```suggestion
```

MikyungKim reviewed (changes requested):
Please find my comments and fix it.

mmyelyn commented on samples/qa-virtualgridlist/src/context/RecordContext.js:
```diff
@@ -0,0 +1,168 @@
+
+export default function recordReducer (state, action) {
+	switch (action.type) {
+		case (ADD_ITEM): {
```
I think this is not a matter of preference. An independent block scope must be created to prevent variables from overlapping.
(https://stackoverflow.com/questions/42480949/what-do-the-curly-braces-do-in-switch-statement-after-case-in-es6)

MikyungKim replied:
I meant ( ) from ADD_ITEM :)

mmyelyn replied:
Ok. I took it as a different situation at first. It's modified

mmyelyn commented:
@MikyungKim Ok, All reviews have been reflected. Also, I accidentally commit the package-lock file, so I removed it.

MikyungKim commented on samples/qa-virtualgridlist/src/components/ImageItem/ImageItemComponent.js:
```diff
@@ -0,0 +1,27 @@
+import {useContext} from 'react';
+import ImageItem from '@enact/sandstone/ImageItem';
```
```suggestion
import ImageItem from '@enact/sandstone/ImageItem';
import {useContext} from 'react';
```

MikyungKim commented on samples/qa-virtualgridlist/src/components/ImageItem/package.json:
```diff
@@ -1,3 +1,3 @@
 {
-  "main": "ImageItem.js"
+  "main": "ImageItemComponent.js"
```
Let's make another name and please rename the folder too.

MikyungKim commented on samples/qa-virtualgridlist/src/components/PanelHeader/PanelHeader.js:
```diff
@@ -22,202 +21,116 @@ import {
+					{!showOverlay && <CheckboxItem onClick={onChangeDirection}>Horizontal</CheckboxItem>}
+				</Cell>
+				<Cell>
+					{/* chnage scroll mode */}
```
typo~
```suggestion
					{/* change scroll mode */}
```

MikyungKim reviewed (approved):
LGTM
